client/network/src/protocol/generic_proto/handler/group.rs

```diff
@@ -461,21 +568,93 @@ impl ProtocolsHandler for NotifsHandler {
 	) -> Poll<
 		ProtocolsHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::OutEvent, Self::Error>
 	> {
+		if let Some(notifications_sink_rx) = &mut self.notifications_sink_rx {
+			'poll_notifs_sink: loop {
+				// Before we poll the notifications sink receiver, check that all the notification
+				// channels are ready to send a message.
```

Comment:
> Is this similar to what you have in mind?

Yes!

> I'd like to better understand how the legacy substream makes one-channel-per-protocol impossible at the moment.

Thinking about it, I think this is not impossible. I have probably gotten confused by a previous iteration.

Should I make the changes now? Splitting into multiple `NotificationsSink` is planned for #5670 anyway.

Reply:
> Should I make the changes now? Splitting into multiple NotificationsSink is planned for #5670 anyway.

I have no preference, but if it seems easier to do together with the removal of the legacy substream, that may be a reason to do it afterwards indeed. It may be helpful though to add notes in the code already at places that are known to be suboptimal right now but will be changed soon, like the coupling of the substreams here.